### PR TITLE
Added a LowVelocityFilter to the input system. 

### PR DESCRIPTION
**Pull request by druiz17**
I moved the LowVelocityFliter that was in MyAvatar.ccp and created an actual Filter that can be used in the input system. This lowVelocity filtering is not enabled by default.

### CODE REVIEW

**Comment by druiz17** [comment]
Test plan
- find the location where this PR is installed. then go to resources->controllers and file the called vive.json
- rename that file to what every you like, then download the file vive-filtering.json (located below).
-move vive-filtering.json into the controllers folder and rename it vive.json
- launch interface, calibrate the pucks and examine the filtering results.



[vive-filtering.zip](https://github.com/highfidelity/hifi/files/1005075/vive-filtering.zip)

**Comment by hifi-gustavo** [comment]
Build console output for commit: add45df4963ec90044f1cde1cb39760ca2dc6501 available at: https://gist.github.com/87cb49474eb5ecdecf30a5a5ba62c157<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-add45df.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-add45df.dmg

**Comment by YuppyPlays** [comment]
Hi there, just curious when you reference Pucks, are the extra Vive Pucks necessary for this QA or is a Vive sufficient enough?

**Comment by druiz17** [comment]
The extra Vive Pucks are necessary.

**Comment by YuppyPlays** [comment]
Ah, I will leave this up to a tester with the extra pucks.

I'll have to order some soon so I can assist in future! :)

**Comment by hifi-gustavo** [comment]
Build console output for commit: 25b1bb0633d3144c77b03ed2fc4bcacfab77df41 available at: https://gist.github.com/87cb49474eb5ecdecf30a5a5ba62c157<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-25b1bb0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-25b1bb0.dmg

**Comment by anshuman64** [comment]
It seems like one of the pucks (my right foot puck) wasn't recognized / wasn't calibrating correctly. I tried restarting SteamVR & Interface but still got these results. Because of this, was not able to determine the effect of the filtering. This was specific to the PR, because when I tried in master everything worked correctly.

Video: https://drive.google.com/file/d/0B_nmInIG19-uYncyaUlDN1NvRmM/view?usp=sharing

**Comment by hifi-gustavo** [comment]
Build console output for commit: e4c79926c0f206c670f8a1a5cfcd3084c8e53fa6 available at: https://gist.github.com/87cb49474eb5ecdecf30a5a5ba62c157<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-e4c7992.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10469-e4c7992.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/66/HighFidelity-Beta-PR10469-ffcc3c8df9ccab8b8f99c5ab0d068b8abbc09ad1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/66/HighFidelity-Beta-PR10469-ffcc3c8df9ccab8b8f99c5ab0d068b8abbc09ad1.dmg

File: libraries/controllers/src/controllers/impl/filters/LowVelocityFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            float rotationFilter = glm::clamp(1.0f - (glm::length(lastPose.getVelocity() / _rotationConstant)), 0.0f, 1.0f);
+            float translationFilter = glm::clamp(1.0f - (glm::length(lastPose.getVelocity() / _translationConstant)), 0.0f, 1.0f);
+            sensorPose.translation = lastPose.getTranslation() * translationFilter + sensorPose.getTranslation() * (1.0f - translationFilter);
+            sensorPose.rotation = safeMix(lastPose.getRotation(), sensorPose.getRotation(), 1.0f - rotationFilter);
+
+            finalPose = sensorPose.transform(sensorToAvatarMat);
+        }
+        _oldPose = finalPose.transform(avatarToSensorMat);
+        return finalPose;
+    }
```
Comment (druiz17) [inline libraries/controllers/src/controllers/impl/filters/LowVelocityFilter.cpp]: Valid means that pose data comes from an object the is successfully tracked by oculus/vive  
Essentially I want to avoid filtering an invalid and a valid pose. Meaning you must have two frames of valid poses to perform the filter. That's is way is set the `_oldPose` to the `finalPose` regardless if the `finalPose` is valid or invalid

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/118/HighFidelity-Beta-PR10469-54d63f515503fdb2016693f030a8db0c461b7447.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/118/HighFidelity-Beta-PR10469-54d63f515503fdb2016693f030a8db0c461b7447.dmg

File: libraries/controllers/src/controllers/impl/filters/LowVelocityFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        const InputCalibrationData calibrationData = userInputMapper->getInputCalibrationData();
+        glm::mat4 sensorToAvatarMat = glm::inverse(calibrationData.avatarMat) * calibrationData.sensorToWorldMat;
+        glm::mat4 avatarToSensorMat = glm::inverse(calibrationData.sensorToWorldMat) * calibrationData.avatarMat;
+        Pose finalPose = newPose;
+        if (finalPose.isValid() && _oldPose.isValid()) {
+            Pose sensorPose = finalPose.transform(avatarToSensorMat);
+            Pose lastPose = _oldPose;
+
+            float rotationFilter = glm::clamp(1.0f - (glm::length(lastPose.getVelocity() / _rotationConstant)), 0.0f, 1.0f);
+            float translationFilter = glm::clamp(1.0f - (glm::length(lastPose.getVelocity() / _translationConstant)), 0.0f, 1.0f);
```
Comment (AndrewMeadows) [inline libraries/controllers/src/controllers/impl/filters/LowVelocityFilter.cpp]: Interesting way to blend between aggressive filtering (lastPose.getVelocity() has very low length) and wide open sampling (lastPose.getVelocity() larger than _translationConstant).  Both the rotational and linear parts depend on the historical linear velocity (you don't use the angular velocity for computing the filter rate, even for the rotational part), but can have independent threshold speeds.

The name "_rotationConstant" is a bit vague.  Its units are m/sec, so it is a "speed scale".

In short, I'm sure it works for some selected values.  If they don't have to be changed very often by non-experts then ok.
